abi-spec/core.go

```diff
@@ -0,0 +1,142 @@
+package abispec
+
+import (
+	"encoding/hex"
+	"encoding/json"
+	"fmt"
+	"math/big"
+	"reflect"
+	"regexp"
+	"strings"
+
+	"github.com/ethereum/go-ethereum/accounts/abi"
+	"github.com/ethereum/go-ethereum/common"
+)
+
+var methodPattern = regexp.MustCompile(`(^[a-zA-Z].*)\((.*)\)`)
+
+var maxSafeInteger = big.NewInt(int64(9007199254740991))
+
+func Encode(method string, paramsJSON string) (string, error) {
+	matches := methodPattern.FindStringSubmatch(method)
+	if len(matches) != 3 {
+		return "", fmt.Errorf("unrecognized method: %s", method)
+	}
+	methodName := matches[1]
+	paramTypesString := strings.TrimSpace(matches[2])
+
+	// value of inputs looks like: `[{ "type": "uint32" },{ "type": "bool" }]`
+	inputs := "["
+	var params []interface{}
+	if len(paramTypesString) > 0 {
+		var paramsRaw []json.RawMessage
+		if err := json.Unmarshal([]byte(paramsJSON), &paramsRaw); err != nil {
+			return "", fmt.Errorf("unable to unmarshal params: %v", err)
+		}
+		types := strings.Split(paramTypesString, ",")
+		if len(paramsRaw) != len(types) {
+			return "", fmt.Errorf("num of param type should equal to num of param value, actual value: %d, %d", len(types), len(paramsRaw))
+		}
+
+		for i, typeName := range types {
+			if i != 0 {
+				inputs += ","
+			}
+			inputs += fmt.Sprintf(`{"type":"%s"}`, typeName)
+
+			param, err := toGoTypeValue(typeName, paramsRaw[i])
+			if err != nil {
+				return "", err
+			}
+			params = append(params, param.Interface())
+		}
+	}
+	inputs += "]"
+
+	inDef := fmt.Sprintf(`[{ "name" : "%s", "type": "function", "inputs": %s}]`, methodName, inputs)
+	inAbi, err := abi.JSON(strings.NewReader(inDef))
+	if err != nil {
+		return "", fmt.Errorf("invalid ABI definition %s, %v", inDef, err)
+	}
+	var result []byte
+	result, err = inAbi.Pack(methodName, params...)
+
+	if err != nil {
+		return "", fmt.Errorf("Pack failed: %v", err)
+	}
+
+	return "0x" + hex.EncodeToString(result), nil
+}
+
+// override result to make it looks like what status-mobile need
```

Comment:
could you please elaborate on what status-mobile needs? and is it possible to change status-mobile so we don't have to override?

Reply:
we need override, otherwise status-mobile need take care of result of json.Marshal, As per the docs: https://golang.org/pkg/encoding/json/#Marshal , 
```
Array and slice values encode as JSON arrays, except that []byte encodes as a base64-encoded string, and a nil slice encodes as the null JSON object.
```